torch/csrc/autograd/python_variable.cpp

```diff
@@ -143,30 +162,70 @@ static int THPVariable_clear(THPVariable *self)
     if (auto grad_acc = torch::autograd::impl::try_get_grad_accumulator(tensor)) {
       grad_acc->pre_hooks().clear();
     }
-    // We must clear the pyobj field in the base C++ Variable, to ensure
-    // that if we attempt to pass the Variable to Python, we don't
-    // attempt to reuse the (now-dead) PyObject.
-    //
-    // One non-obvious consequence of this: if you have a tensor x, you
-    // take its id(), and then you let it become dead in Python, if you
-    // get another reference to the tensor in Python later (because you
-    // passed it from C++ to Python), you'll get a *different* id() the
-    // second time around.  So you better make sure that if you're using
-    // id() to keep track of Tensors, you better make sure their Python
-    // objects stay live, buster!  See
-    // https://github.com/pytorch/pytorch/issues/22884 for an example of
-    // this actually showing up.
-    set_pyobj(self->cdata, nullptr);
-  }
-  self->cdata.reset();
+    set_pyobj(tensor, nullptr);
+    tensor.unsafeGetTensorImpl()->set_owns_pyobj(false);
+  }
+  self->cdata = MaybeOwned<Variable>();
   return 0;
 }
 
+// returns true if successfully rezzed; if so, cancel the
+// rest of deallocation
+static bool THPVariable_tryResurrect(THPVariable* self) {
+  const auto& tensor = THPVariable_Unpack(self);
+
+  // Check if there are other C++ owners
+  if (tensor.use_count() <= 1) {
+    return false;
+  }
+
+  // There are other C++ owners of the tensor.  Flip ownership
+  // so that C++ owns this Python object, and cancel deallocation.
+  TORCH_INTERNAL_ASSERT(!tensor.unsafeGetTensorImpl()->owns_pyobj());
+
+  tensor.unsafeGetTensorImpl()->set_owns_pyobj(true);
+
+  // Resurrect the Python object.  This is something CPython does
+  // internally occasionally, see
+  // https://github.com/python/cpython/blob/b98eba5bc2ffbe7a0ed49d540ebc4f756ae61985/Objects/object.c#L248-L259
+  // so we just copy the pattern here.  Note that we don't have to worry
+  // about saving and restoring the refcount (as the quoted code does)
+  // because we actually DO need to reset the refcount to one here, we
+  // can't assume that some other code has taken care of it.
+  // NB: this will overreport _Py_RefTotal but based on inspection of object.c
+  // there is no way to avoid this
+  _Py_NewReference((PyObject*)self);
```

Comment:
I wonder if this should just be:

```
#ifdef Py_TRACE_REFS
    _Py_AddToAllObjects(op, 1);
#endif
   Py_SET_REFCNT(self, 1);  // or Py_INCREF(self); ???
```

The other bits of `_Py_NewReference` seem... unhelpful. I *think* `Py_TRACE_REFS` should be available via Python.h header.